### PR TITLE
refactor: Remove MakeUnique<T>()

### PR DESCRIPTION
**Pull request by fanquake**
Since requiring C++17, this is just pointless abstraction. I think we should just "tear the band-aid off" and remove it. Similar to the changes happening in #21366.

Also, having a comment saying this is deprecated doesn't prevent it's usage in new code. i.e : https://github.com/bitcoin/bitcoin/pull/20946#discussion_r561949731.

The repository is fairly quiet at the moment, so any potential complaints about having to rebase should be minimal. Might as well get this over and done with.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Seems fine, but let's wait for the list of conflicts. Two minor style nits (feel free to ignore):

* I think `git rm` is preferred over `rm -f` because it will fail if the file is not tracked by git
* I think a third commit can be added to bump the copyright header of the files. For some files this will be the only change this year and to avoid touching them twice for a trivial change, both changes could be done in this pull.

File: doc/developer-notes.md [hunk 1]
```diff
@@ -595,10 +595,9 @@ Common misconceptions are clarified in those sections:
 
   - *Rationale*: This avoids memory and resource leaks, and ensures exception safety.
 
-- Use `MakeUnique()` to construct objects owned by `unique_ptr`s.
+- Use `std::make_unique` to construct objects owned by `unique_ptr`s.
```
Comment (jnewbery) [inline doc/developer-notes.md:598]: I don't think this needs to go in our developer notes at all. It's general best practice when writing modern c++ that you should use std::make_unique to construct unique pointers. We don't need to document all those best practices in our project.

**Review by jnewbery — COMMENTED** [review]
Concept ACK, although I agree with Marco that we should wait for drahtbot to tell us how many PRs this conflicts with, and hold back if any of those are important and close to being merged.

Will this silently conflict with PRs that introduce `MakeUnique()` calls? Is there a way to find those and force a CI rerun?

**Comment by practicalswift** [comment]
Concept ACK: nice cleanup!

**Comment by fanquake** [comment]
Looks like the list of potential conflicts is fairly small, only 13. With some of those already being rebased regularly.
There's one with any ACKs, #20228, which has a single ACK on the latest commit hash.
There's at least one, #21365, which introduces new `MakeUnique` usage, so [I've commented there](https://github.com/bitcoin/bitcoin/pull/21365/commits/d5a0dce8fe40153b07ec6e882e8b1568b05e3905#r592082640).

I've taken the `git rm` suggestion, and dropped `std::make_unique` from the developer docs.

File: doc/developer-notes.md (hunk 1 above)
Comment (fanquake) [reply doc/developer-notes.md]: This has now been removed.

**Comment by jnewbery** [comment]
utACK 1a6323bdbe20bdb7b1c907d8fa0333ffa88b21ff

Change looks correct. No comment on when this should be merged. I'll defer to the wisdom of the maintainers :mage: :mage: :mage:

**Comment by practicalswift** [comment]
cr ACK 1a6323bdbe20bdb7b1c907d8fa0333ffa88b21ff: patch looks correct

**Review by glozow — COMMENTED** [review]
ACK https://github.com/bitcoin/bitcoin/commit/1a6323bdbe20bdb7b1c907d8fa0333ffa88b21ff looks correct

> Will this silently conflict with PRs that introduce MakeUnique() calls? Is there a way to find those and force a CI rerun?

Sounds like fanquake has been :eyes:ing the `MakeUnique`s for a while. But generally, it sounds like we have a lot of things we want to get rid of. If we want to soak the bandaid a little first, could write a linter for them, add the linter to CI and give it some time to catch new occurrences, then land the removal PR. Otherwise you basically have to flush the pipeline, either by searching for the silent conflicts manually or re-running a bunch of CI?

**Comment by fanquake** [comment]
> could write a linter for them, add the linter to CI and give it some time to catch new occurrences, then land the removal PR.

Instead of a new linter, I wrote a few lines of Python to call the GitHub API, and list me all the PRs that introduce new `MakeUnique<>` usage. I've now left comments on all the relevant PRs advising that `std::make_unique` should be used. i.e: [here](https://github.com/bitcoin/bitcoin/pull/21186#discussion_r592834789), [here](https://github.com/bitcoin/bitcoin/pull/18261#discussion_r592830262) and [here](https://github.com/bitcoin/bitcoin/pull/21365/#discussion_r592082640).

**Comment by ajtowns** [comment]
ACK 1a6323bdbe20bdb7b1c907d8fa0333ffa88b21ff -- code review only

Rescanning open PRs after this is merged should be enough for CI to catch everything I think? Maybe close/reopen any PRs that still hit so CI gets rerun and errors on them?

**Pull request merged**